src/java/org/apache/cassandra/utils/concurrent/OpOrder.java

```diff
@@ -318,21 +321,21 @@ public boolean isBlocking()
             return isBlocking;
         }
 
-        /**
-         * register to be signalled when a barrier waiting on us is, or maybe, blocking general progress,
-         * so we should try more aggressively to progress
-         */
-        public WaitQueue.Signal isBlockingSignal()
+        public void notifyIfBlocking(WaitQueue.Signal signal)
         {
-            return isBlockingSignal.register();
+            if (blocking == null)
+                blockingUpdater.compareAndSet(this, null, new ConcurrentLinkedQueue<>());
```

Comment:
As I can see it is set once, why not to set it for the field initially?

```
private volatile ConcurrentLinkedQueue<WaitQueue.Signal> blocking; // signal to wait on to indicate isBlocking is true
```
-> 
```
private final ConcurrentLinkedQueue<WaitQueue.Signal> blocking = new ConcurrentLinkedQueue<>();
```

And why did we migrate from `WaitingQueue`?

Reply:
> why not to set it for the field initially?

To save memory

> And why did we migrate from WaitingQueue?

If you interpret the remaining behavioural changes you can probably see why, but in essence it boils down to `WaitQueue.any` implementation being incompatible with the Simulator and mocking its implementation

Reply:
Ok, assuming `notifyIfBlocking` is not called for every instance of `Group`, that makes sense.

> If you interpret the remaining behavioural changes you can probably see why

That's the problem for me, I'm not really clear on all these structural changes. I understand them partially and that's why I asked :)